arch/risc-v/src/common/espressif/esp_qencoder.c

```diff
@@ -0,0 +1,593 @@
+{
+  struct esp_lowerhalf_s *priv = (struct esp_lowerhalf_s *)lower;
+  irqstate_t flags;
+  uint32_t regval;
+
+  sninfo("Resetting position to zero\n");
+  DEBUGASSERT(lower && priv->inuse);
+
+  /* Reset this Pulse Counter Unit. */
+
+  flags = spin_lock_irqsave(&priv->lock);
+
+  /* Reset RST bit and clear RST bit to enable counting again */
+
+  pcnt_ll_clear_count(priv->dev, priv->config->pcntid);
+
+  priv->position = 0;
+
+  spin_unlock_irqrestore(&priv->lock, flags);
+  return OK;
+}
+
+/****************************************************************************
+ * Name: esp_setindex
+ *
+ * Description:
+ *   Set the index pin position
+ *
+ ****************************************************************************/
+
+static int esp_setindex(struct qe_lowerhalf_s *lower, uint32_t pos)
+{
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Name: esp_ioctl
+ *
+ * Description:
+ *   Lower-half logic may support platform-specific ioctl commands
+ *
+ ****************************************************************************/
+
+static int esp_ioctl(struct qe_lowerhalf_s *lower, int cmd,
+                       unsigned long arg)
+{
+  /* No ioctl commands supported */
+
+  /* TODO add an IOCTL to control the encoder pulse count prescaler */
+
+  return -ENOTTY;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: esp_qeinitialize
+ *
+ * Description:
+ *   Initialize a quadrature encoder interface.  This function must be
+ *   called from board-specific logic.
+ *
+ * Input Parameters:
+ *   devpath - The full path to the driver to register. E.g., "/dev/qe0"
+ *   pcnt    - The PCNT number to used.  'pcnt' must be an element of
+ *             {0,1,2,3}
+ *
+ * Returned Value:
+ *   Zero on success; A negated errno value is returned on failure.
+ *
+ ****************************************************************************/
+
+int esp_qeinitialize(const char *devpath, int pcnt)
+{
+  struct esp_lowerhalf_s *priv;
+  int ret;
+
+  /* Find the pre-allocated timer state structure corresponding to this
+   * timer
+   */
+
+  priv = esp_pcnt2lower(pcnt);
+  if (priv == NULL)
+    {
+      snerr("ERROR: PCNT%d support not configured\n", pcnt);
+      return -ENXIO;
+    }
+
+  /* Make sure that it is available */
+
+  if (priv->inuse)
+    {
+      snerr("ERROR: PCNT%d is in-use\n", pcnt);
+      return -EBUSY;
+    }
+
+  /* Register the upper-half driver */
+
+  ret = qe_register(devpath, (struct qe_lowerhalf_s *)priv);
+  if (ret < 0)
+    {
+      snerr("ERROR: qe_register failed: %d\n", ret);
+      return ret;
+    }
+
+  /* Make sure that the PCNT is in the shutdown state */
+
+  esp_shutdown((struct qe_lowerhalf_s *)priv);
```

Comment:
I recommend checking if the peripheral module is disabled by default during startup. If it is, just set the `position`to 0 and that's it.

Reply:
I now enable clock and reset the PCNT peripheral during initialization instead of during setup as it affects all units, so I could have inadvertently reset the state of other units. Please, check if you agree.